File: packages/dataregistry/src/dataregistry.ts
import { Dataset } from './dataset';
import { JSONValue, Token } from '@lumino/coreutils';
import { Signal } from '@lumino/signaling';

const DatasetStore: { [key: string]: Dataset<any, any>[] } = {};
const CommandsStore: { [key: string]: Set<string> } = {};
const SignalsStore: { [key: string]: Signal<any, any> } = {};

export interface IDataRegistry {
  registerDataset<T extends JSONValue, U extends JSONValue>(
    dataset: Dataset<T, U>
  ): void;

  updateDataset<T extends JSONValue, U extends JSONValue>(
    dataset: Dataset<T, U>
  ): void;

  getDataset<T extends JSONValue, U extends JSONValue>(
    id: string,
    version?: string
  ): Dataset<T, U>;

  getDatasetSignal<T extends JSONValue, U extends JSONValue>(
    id: string
  ): Signal<any, Dataset<T, U>>;

  hasDataset<T extends JSONValue, U extends JSONValue>(
    id: string,
    version?: string
  ): boolean;

  registerCommand(
    commandId: string,
    abstractDataType: string,
    serializationType: string,
    storageType: string
  ): void;

  getCommands(
    abstractDataType: string,
    serializationType: string,
    storageType: string
  ): Set<string> | [];

  queryDataset(
    abstractDataType?: string,
    serializationType?: string,
    storageType?: string
  ): Dataset<any, any>[] | [];

  readonly datasetAdded: Signal<any, Dataset<any, any>>;
  readonly datasetUpdated: Signal<any, Dataset<any, any>>;
  readonly commandAdded: Signal<any, String>;
}

class Registry implements IDataRegistry {
  readonly datasetAdded: Signal<any, Dataset<any, any>>;
  readonly datasetUpdated: Signal<any, Dataset<any, any>>;
  readonly commandAdded: Signal<any, String>;

  constructor() {
    this.datasetAdded = new Signal(this);
    this.datasetUpdated = new Signal(this);
    this.commandAdded = new Signal(this);
  }

  /**
   * Registers a dataset, throws an exception if dataset already exists.
   * @param dataset The dataset to register
   */
  registerDataset<T extends JSONValue, U extends JSONValue>(
    dataset: Dataset<T, U>
  ) {
    _registerDataset(dataset);
    this.datasetAdded.emit(dataset);
  }

  /**
   * Updates a registered dataset, bumps up the version
   * @param dataset The dataset to update
   */
  updateDataset<T extends JSONValue, U extends JSONValue>(
    dataset: Dataset<T, U>
  ) {
    const registeredDataset = this.getDataset(dataset.id);
    if (registeredDataset) {
      if (registeredDataset.abstractDataType !== dataset.abstractDataType) {
        throw new Error(
          `Abstract data type "${dataset.abstractDataType}" doesn't match "${registeredDataset.abstractDataType}"`
        );
      }
      if (registeredDataset.serializationType !== dataset.serializationType) {
        throw new Error(
          `Serialization type "${dataset.serializationType}" doesn't match "${registeredDataset.serializationType}"`
        );
      }
      if (registeredDataset.storageType !== dataset.storageType) {
        throw new Error(
          `Storage type "${dataset.storageType}" doesn't match "${registeredDataset.storageType}"`
        );
      }
      _registerDataset(dataset);
      this.getDatasetSignal(dataset.id).emit(dataset);
      this.datasetUpdated.emit(dataset);
    }
  }

  /**
   * Returns last registered version of dataset if no version is passed
   * @param id unique id dataset was registered with
   * @param version optional, specific dataset version
   */
  getDataset<T extends JSONValue, U extends JSONValue>(
    id: string,
    version?: string
  ): Dataset<T, U> {
    const datasets = DatasetStore[id];
    if (!datasets) {
      throw new Error(`No dataset with id: ${id} exists.`);
    } else if (version) {
      const dataset = datasets.find((dataset) => dataset.version === version);
      if (dataset) {
        return dataset;
      } else {
        throw new Error(
          `No dataset with id: ${id}, version: ${version} exists.`
        );
      }
    }
    return datasets[datasets.length - 1];
  }

  /**
   * Returns dataset signal for subscribing to changes in dataset
   * @param id unique id used to register the dataset
   */
  getDatasetSignal<T extends JSONValue, U extends JSONValue>(
    id: string
  ): Signal<any, Dataset<T, U>> {
    if (this.hasDataset(id)) {
      return SignalsStore[id];
    } else {
      throw new Error(`No dataset registered with id: ${id}`);
    }
  }

  /**
   * Returns true if dataset exists, false otherwise
   * @param id unique id that was used to register the dataset
   * @param version version that dataset was registered with (optional, defaults to 0)
   */
  hasDataset<T extends JSONValue, U extends JSONValue>(
    id: string,
    version?: string
  ): boolean {
    const datasets = DatasetStore[id];
    return (
      datasets &&
      (version
        ? datasets.findIndex((ds) => ds.version === version) !== -1
        : true)
    );
  }

  /**
   * Registers a command for an abstract data type, serialization type,
   * and storage type. This can be used to get a list of commands that
   * that a dataset can support.
   * @param commandId unique id of the command registered with the command registry
   * @param abstractDataType
   * @param serializationType
   * @param storageType
   */
  registerCommand(
    commandId: string,
    abstractDataType: string,
    serializationType: string,
    storageType: string
  ): void {
    const key = `${abstractDataType}:${serializationType}:${storageType}`;
    const actions = CommandsStore[key] || new Set();
    CommandsStore[key] = actions.add(commandId);
    this.commandAdded.emit(commandId);
  }

  /**
   * Get list of registered commands
   * @param abstractDataType
   * @param serializationType
   * @param storageType
   */
  getCommands(
    abstractDataType: string,
    serializationType: string,
    storageType: string
  ): Set<string> | [] {
    return (
      CommandsStore[
        `${abstractDataType}:${serializationType}:${storageType}`
      ] || []
    );
  }

  /**
   * Returns datasets that match the passed abstract data type,
   * serialization type, and storage type.
   * @param abstractDataType
   * @param serializationType
   * @param storageType
   */
  queryDataset(
    abstractDataType?: string,
    serializationType?: string,
    storageType?: string
  ): Dataset<any, any>[] | [] {
    const datasets = [];
    for (const id in DatasetStore) {
      const versions = DatasetStore[id];
      const dataset = versions[versions.length - 1];
      let include = true;
      if (abstractDataType) {
        include = dataset.abstractDataType === abstractDataType;
      }
      if (serializationType) {
        include = include && dataset.serializationType === serializationType;
      }
      if (storageType) {
        include = include && dataset.storageType === storageType;
      }
      if (include) {
        datasets.push(dataset);
      }
    }
    return datasets;
  }
}

const registry = new Registry();

console.log('Created a new registry');

function _registerDataset<T extends JSONValue, U extends JSONValue>(
  dataset: Dataset<T, U>
) {
  if (registry.hasDataset<T, U>(dataset.id, dataset.version)) {
    throw new Error(
      `Dataset with id: ${dataset.id}, version: ${dataset.version} already exists.`
    );
  } else {
    if (!DatasetStore[dataset.id]) {
      DatasetStore[dataset.id] = [dataset];
      SignalsStore[dataset.id] = new Signal<any, Dataset<T, U>>(registry);
    } else {
      DatasetStore[dataset.id] = [...DatasetStore[dataset.id], dataset];
    }
  }
}

export const IDataRegistry = new Token<IDataRegistry>(
  '@jupyterlab/dataregistry:IDataRegistry'
);

export default registry;

File: jupyterlab_dataregistry/dataregistry.py

from argparse import ArgumentError
import os
from typing import Optional
from .command_store import CommandStore
from .dataset import Dataset

from .dataset_store import DatasetStore


class AlreadyExistsError(Exception):
    def __init__(self, id: str, version: Optional[str] = None):
        self.id = id
        self.version = version

    def __str__(self):
        return f"Dataset with id: {self.id} and version: {self.version} already exists."


class NotFoundError(Exception):
    def __init__(self, id: str, version: Optional[str] = None):
        self.id = id
        self.version = version

    def __str__(self):
        return f"Dataset with id: {self.id} and version: {self.version} not found."


class TypeMismatchError(Exception):
    def __init__(self, type_name: str, actual: str, expected: str):
        self.type_name = type_name
        self.actual = actual
        self.expected = expected

    def __str__(self):
        return f"{self.actual} passed for {self.type_name}, but {self.expected} expected."


class DataRegistry():

    def __init__(self, store_path: str):
        base_path = os.path.join(store_path, ".dataregistry")

        self.__dataset_store = DatasetStore(os.path.join(base_path, "datasets"))
        self.__command_store = CommandStore(os.path.join(base_path, "commands"))

    
    def register_dataset(self, dataset: Dataset):
        dataset_store = self.__dataset_store
        id = dataset.id
        version = dataset.version
        if dataset_store.has_dataset(id, version):
            raise AlreadyExistsError(id, version)
        else:
            dataset_store.save_dataset(dataset)


    def update_dataset(self, dataset: Dataset):
        dataset_store = self.__dataset_store
        id = dataset.id
        version = dataset.version
        
        registered_dataset = dataset_store.load_dataset(id)
        if registered_dataset.abstract_data_type != dataset.abstract_data_type:
            raise TypeMismatchError(
                "abstractDataType",
                dataset.abstract_data_type,
                registered_dataset.abstract_data_type
            )
        
        if registered_dataset.serialization_type != dataset.serialization_type:
            raise TypeMismatchError(
                "serializationType",
                dataset.serialization_type,
                registered_dataset.serialization_type
            )
        
        if registered_dataset.storage_type != dataset.storage_type:
            raise TypeMismatchError(
                "storageType",
                dataset.storage_type,
                registered_dataset.storage_type
            )
        
        if dataset_store.dataset_exists(id, version):
            raise AlreadyExistsError(id, version)
        else:
            dataset_store.save_dataset(dataset)
        

    def get_dataset(self, id, version = None):
        dataset_store = self.__dataset_store
        if dataset_store.has_dataset(id, version):
            dataset = dataset_store.load_dataset(id, version)
        else:
            raise NotFoundError(id, version)

        return dataset


    def query_dataset(
        self, 
        abstract_data_type: str = None, 
        serialization_type: str = None, 
        storage_type: str = None
    ):
        datasets = []
        for dataset in self.__dataset_store.load_all_datasets():
            include = True
            if abstract_data_type:
                include = dataset.abstract_data_type == abstract_data_type

            if serialization_type:
                include = include and dataset.serialization_type == serialization_type

            if storage_type:
                include = include and dataset.storage_type == storage_type

            if include:
                datasets.append(dataset)

        return datasets


    def has_dataset(
        self,
        id: str,
        version: str = None
    ) -> Dataset:  
        dataset = self.__dataset_store.has_dataset(id, version)
        return dataset


    def register_command(
        self,
        command_id: str,
        abstract_data_type: str, 
        serialization_type: str, 
        storage_type: str
    ):
        self.__command_store.add_command(
            command_id, 
            abstract_data_type,
            serialization_type,
            storage_type
        )       


    def get_commands(
        self,
        abstract_data_type: str, 
        serialization_type: str, 
        storage_type: str
    ):
        commands = self.__command_store.get_commands(
            abstract_data_type,
            serialization_type,
            storage_type
        )

        return commands
    
    